File: .agent/mcp/mcp-project-memory/src/inspector.py
"""Debug/inspect queries for AI visibility into raw data."""

import json
from dataclasses import asdict

from src.db import Database
from src.memory.stores import BuildMetaStore, LinkStore, MemoryStore


class Inspector:
    """Debug/inspect queries for AI visibility into raw data."""

    def __init__(
        self,
        db: Database,
        memory_store: MemoryStore,
        link_store: LinkStore,
        build_meta_store: BuildMetaStore,
    ):
        self._db = db
        self._memories = memory_store
        self._links = link_store
        self._build_meta = build_meta_store

    def inspect(self, query: str) -> str:
        """Route an inspect query to the appropriate handler."""
        parts = query.strip().lower().split()
        if not parts:
            return self._help()

        cmd = parts[0]
        args = parts[1:]

        handlers = {
            "tables": self._tables,
            "memories": self._all_memories,
            "memory": self._single_memory,
            "links": self._all_links,
            "builds": self._builds,
            "stats": self._stats,
            "schema": self._schema,
            "fts": self._fts_health,
            "help": self._help,
        }

        handler = handlers.get(cmd)
        if handler is None:
            return f"Unknown inspect command: {cmd}\n\n{self._help()}"

        return handler(*args)

    def _help(self, *_: str) -> str:
        return (
            "Inspect commands:\n"
            "  tables            — List all tables\n"
            "  memories          — Show all memories\n"
            "  memory <id>       — Show a specific memory with links\n"
            "  links             — Show all links\n"
            "  builds            — Show build history\n"
            "  stats             — Aggregate statistics\n"
            "  schema            — Show table schemas\n"
            "  fts               — FTS5 index health check\n"
            "  help              — This message"
        )

    def _tables(self, *_: str) -> str:
        rows = self._db.query(
            "SELECT name, type FROM sqlite_master WHERE type IN ('table', 'view') ORDER BY name"
        )
        return json.dumps([{"name": r["name"], "type": r["type"]} for r in rows], indent=2)

    def _all_memories(self, *_: str) -> str:
        memories = self._memories.list_all(active_only=False, limit=200)
        return json.dumps([m.to_dict() for m in memories], indent=2)

    def _single_memory(self, *args: str) -> str:
        if not args:
            return "Usage: memory <id>"
        memory_id = args[0]
        memory = self._memories.get(memory_id)
        if memory is None:
            return f"Memory {memory_id} not found"
        links = self._links.get_links_for(memory_id)
        return json.dumps({
            "memory": memory.to_dict(),
            "links": [asdict(lnk) for lnk in links],
        }, indent=2)

    def _all_links(self, *_: str) -> str:
        links = self._links.list_all(limit=200)
        return json.dumps([asdict(lnk) for lnk in links], indent=2)

    def _builds(self, *_: str) -> str:
        builds = self._build_meta.list_all()
        return json.dumps([asdict(b) for b in builds], indent=2)

    def _stats(self, *_: str) -> str:
        stats = self._memories.stats()
        last_build = self._build_meta.get_last()
        stats["last_build"] = asdict(last_build) if last_build else None
        return json.dumps(stats, indent=2)

    def _schema(self, *_: str) -> str:
        rows = self._db.query(
            "SELECT sql FROM sqlite_master WHERE sql IS NOT NULL ORDER BY name"
        )
        return "\n\n".join(r["sql"] for r in rows)

    def _fts_health(self, *_: str) -> str:
        try:
            result = self._db.query_one(
                "SELECT COUNT(*) as c FROM memories_fts"
            )
            memory_count = self._memories.count(active_only=False)
            fts_count = result["c"]
            return json.dumps({
                "fts_rows": fts_count,
                "memory_rows": memory_count,
                "in_sync": fts_count == memory_count,
            }, indent=2)
        except Exception as e:
            return json.dumps({"error": str(e)}, indent=2)

File: .agent/mcp/mcp-project-memory/internal/inspector/inspector.go
// Package inspector provides debug/inspect commands for AI visibility
// into raw memory data. Depends on domain interfaces only.
package inspector

import (
	"encoding/json"
	"fmt"
	"strings"

	"github.com/dotai/mcp-project-memory/internal/domain"
)

// Inspector routes debug queries to the appropriate handler.
type Inspector struct {
	memReader domain.MemoryReader
	links     domain.LinkReader
	builds    domain.BuildMetaStore
	dbMgr     domain.DatabaseManager
	rawDB     RawQuerier
}

// RawQuerier provides raw SQL access for schema/table inspection.
type RawQuerier interface {
	Query(query string, args ...any) ([]map[string]any, error)
}

// New creates an Inspector.
func New(
	memReader domain.MemoryReader,
	links domain.LinkReader,
	builds domain.BuildMetaStore,
	dbMgr domain.DatabaseManager,
	rawDB RawQuerier,
) *Inspector {
	return &Inspector{
		memReader: memReader,
		links:     links,
		builds:    builds,
		dbMgr:     dbMgr,
		rawDB:     rawDB,
	}
}

// Inspect routes a query string to the appropriate handler.
func (ins *Inspector) Inspect(query string) string {
	parts := strings.Fields(strings.ToLower(strings.TrimSpace(query)))
	if len(parts) == 0 {
		return ins.help()
	}

	cmd := parts[0]
	args := parts[1:]

	switch cmd {
	case "tables":
		return ins.tables()
	case "memories":
		return ins.allMemories()
	case "memory":
		return ins.singleMemory(args)
	case "links":
		return ins.allLinks()
	case "builds":
		return ins.buildHistory()
	case "stats":
		return ins.stats()
	case "schema":
		return ins.schema()
	case "fts":
		return ins.ftsHealth()
	case "help":
		return ins.help()
	default:
		return fmt.Sprintf("Unknown inspect command: %s\n\n%s", cmd, ins.help())
	}
}

func (ins *Inspector) help() string {
	return `Inspect commands:
  tables            — List all tables
  memories          — Show all memories
  memory <id>       — Show a specific memory with links
  links             — Show all links
  builds            — Show build history
  stats             — Aggregate statistics
  schema            — Show table schemas
  fts               — FTS5 index health check
  help              — This message`
}

func (ins *Inspector) tables() string {
	if ins.rawDB == nil {
		return `{"error": "raw query not available"}`
	}
	rows, err := ins.rawDB.Query("SELECT name, type FROM sqlite_master WHERE type IN ('table', 'view') ORDER BY name")
	if err != nil {
		return fmt.Sprintf(`{"error": %q}`, err.Error())
	}
	data, _ := json.MarshalIndent(rows, "", "  ")
	return string(data)
}

func (ins *Inspector) allMemories() string {
	memories, err := ins.memReader.ListAll(false, 200)
	if err != nil {
		return fmt.Sprintf(`{"error": %q}`, err.Error())
	}
	var dicts []map[string]any
	for _, m := range memories {
		dicts = append(dicts, m.ToDict())
	}
	data, _ := json.MarshalIndent(dicts, "", "  ")
	return string(data)
}

func (ins *Inspector) singleMemory(args []string) string {
	if len(args) == 0 {
		return "Usage: memory <id>"
	}
	m, err := ins.memReader.Get(args[0])
	if err != nil {
		return fmt.Sprintf(`{"error": %q}`, err.Error())
	}
	if m == nil {
		return fmt.Sprintf("Memory %s not found", args[0])
	}
	links, _ := ins.links.GetLinksFor(args[0])
	result := map[string]any{
		"memory": m.ToDict(),
		"links":  links,
	}
	data, _ := json.MarshalIndent(result, "", "  ")
	return string(data)
}

func (ins *Inspector) allLinks() string {
	links, err := ins.links.ListAll(200)
	if err != nil {
		return fmt.Sprintf(`{"error": %q}`, err.Error())
	}
	data, _ := json.MarshalIndent(links, "", "  ")
	return string(data)
}

func (ins *Inspector) buildHistory() string {
	builds, err := ins.builds.ListBuilds(20)
	if err != nil {
		return fmt.Sprintf(`{"error": %q}`, err.Error())
	}
	data, _ := json.MarshalIndent(builds, "", "  ")
	return string(data)
}

func (ins *Inspector) stats() string {
	stats, err := ins.memReader.Stats()
	if err != nil {
		return fmt.Sprintf(`{"error": %q}`, err.Error())
	}
	lastBuild, _ := ins.builds.GetLast()
	stats["last_build"] = lastBuild
	data, _ := json.MarshalIndent(stats, "", "  ")
	return string(data)
}

func (ins *Inspector) schema() string {
	if ins.rawDB == nil {
		return `{"error": "raw query not available"}`
	}
	rows, err := ins.rawDB.Query("SELECT sql FROM sqlite_master WHERE sql IS NOT NULL ORDER BY name")
	if err != nil {
		return fmt.Sprintf(`{"error": %q}`, err.Error())
	}
	var stmts []string
	for _, r := range rows {
		if s, ok := r["sql"].(string); ok {
			stmts = append(stmts, s)
		}
	}
	return strings.Join(stmts, "\n\n")
}

func (ins *Inspector) ftsHealth() string {
	if ins.rawDB == nil {
		return `{"error": "raw query not available"}`
	}
	ftsRows, err := ins.rawDB.Query("SELECT COUNT(*) as c FROM memories_fts")
	if err != nil {
		return fmt.Sprintf(`{"error": %q}`, err.Error())
	}
	var ftsCount int
	if len(ftsRows) > 0 {
		if c, ok := ftsRows[0]["c"].(int64); ok {
			ftsCount = int(c)
		}
	}
	memCount, _ := ins.memReader.Count(false)
	result := map[string]any{
		"fts_rows":    ftsCount,
		"memory_rows": memCount,
		"in_sync":     ftsCount == memCount,
	}
	data, _ := json.MarshalIndent(result, "", "  ")
	return string(data)
}
